### PR TITLE
Varieties v0.3

### PR DESCRIPTION
**Pull request by mahrud**
This PR contains work throughout Spring and Summer 2024, including at the workshop at Utah. Complexes of sheaves are added, but are currently disabled until the Complexes package can be preloaded. Lots of work remains.

cc: @kellerlv @ggsmith @Devlin-Mallory @ritvikrr @johndcobb

### CODE REVIEW

**Comment by mahrud** [comment]
I'm not sure if Mike has forgotten about https://github.com/mikestillman/M2/pull/38, but this PR also includes those changes.

**Comment by mahrud** [comment]
@d-torrance any idea what's happening here? I don't think this PR changes anything related to this ...
```m2
--*- compilation -*-

i1 : input "../../../../../Macaulay2/tests/normal/alarm.m2"

ii2 : -- no-check-flag (see https://github.com/Macaulay2/M2/issues/1392)
      -- test that alarms work
      
      time try ( alarm(1); while true do 1 ) else true
 -- used 1.13874s (cpu); 0.999088s (thread); 0s (gc)

oo2 = true

ii3 : 
      R = QQ[vars(0..24)]

oo3 = R

oo3 : PolynomialRing

ii4 : f = () -> (alarm 4; try res coker vars R else "ran out of time")

oo4 = f

oo4 : FunctionClosure

ii5 : time f()


 *** out of memory trying to allocate 128 bytes, exiting ***
Command exited with non-zero status 1
16.95user 4.55system 0:11.87elapsed 181%CPU (0avgtext+0avgdata 477640maxresident)k
0inputs+8outputs (0major+13659minor)pagefaults 0swaps
```

**Comment by d-torrance** [comment]
Not sure.  Apparently this test has been a problem before since we actually skip it when when running `check Core` (see #1867).

File: M2/Macaulay2/packages/Varieties.m2
Comment (Devlin-Mallory) [inline M2/Macaulay2/packages/Varieties.m2:181]: perhaps the complaint should read "should be homogeneous and generated in degree 1"

**Comment by d-torrance** [comment]
@mahrud - Do you think it's worth applying the `SumsOfSquares` test fix on `development` separately while we wait for this PR to be review?

**Comment by mahrud** [comment]
> Do you think it's worth applying the `SumsOfSquares` test fix on `development` separately while we wait for this PR to be review?

I think you meant to comment this on #3519, where bbc9f09 is, no? I think it's good to keep it there as extra incentive for reviewing it.

**Comment by d-torrance** [comment]
Sure -- it looks like this PR is on top of that one, so either way.  :)  Sounds good!

File: M2/Macaulay2/m2/modules2.m2
Comment (ggsmith) [inline M2/Macaulay2/m2/modules2.m2:355]: Are the ``pullback`` and ``pushout`` documented or tested anywhere?

File: M2/Macaulay2/packages/Book3264Examples.m2
Comment (ggsmith) [inline M2/Macaulay2/packages/Book3264Examples.m2:67]: This appears to depend on other changes to ``basis``, which I am personally not yet sold on.

File: M2/Macaulay2/packages/Complexes/CartanEilenberg.m2
Comment (ggsmith) [inline M2/Macaulay2/packages/Complexes/CartanEilenberg.m2:418]: I believe that the ``SpectralSequences`` package has enjoy a major revision which I suspect is incompatible with this code.

File: M2/Macaulay2/m2/basis.m2
Comment (ggsmith) [inline M2/Macaulay2/m2/basis.m2:240]: Aren't these many changes to ``basis`` already part of an independent pull request?

File: M2/Macaulay2/m2/matrix2.m2
Comment (ggsmith) [inline M2/Macaulay2/m2/matrix2.m2:343]: Can we make the (breaking?) changes to quotients of ``Matrices`` a separate pull request?  I was unable to find the associated documentation and tests.

File: M2/Macaulay2/packages/Isomorphism.m2
Comment (ggsmith) [inline M2/Macaulay2/packages/Isomorphism.m2:196]: - [x] good catch

File: M2/Macaulay2/packages/Complexes/ChainComplex.m2
Comment (ggsmith) [inline M2/Macaulay2/packages/Complexes/ChainComplex.m2:1165]: - [x] good idea

File: M2/Macaulay2/packages/Complexes/ChainComplexMap.m2
Comment (ggsmith) [inline M2/Macaulay2/packages/Complexes/ChainComplexMap.m2:104]: I think that this change is already part of a pull request associated to ``Complexes``.

File: M2/Macaulay2/packages/NormalToricVarieties/Divisors.m2
Comment (ggsmith) [inline M2/Macaulay2/packages/NormalToricVarieties/Divisors.m2:180]: By definition, the nef cone lives in the Neron--Severi group (which for a toric variety is canonical isomorphic to the Picard group).  Does returning something in the class group make mathematical sense?

**Review by ggsmith — CHANGES_REQUESTED** [review]
Can this pull request be decoupled from the changes to ``basis``?  Can the changes to quotient modules and matrices to extracted?

File: M2/Macaulay2/packages/NormalToricVarieties/Divisors.m2
Comment (mahrud) [reply M2/Macaulay2/packages/NormalToricVarieties/Divisors.m2:180]: Sorry, I don't follow. The only change here is caching of `nefGenerators`. What change are you referring to?

File: M2/Macaulay2/packages/Complexes/ChainComplexMap.m2
Comment (mahrud) [reply M2/Macaulay2/packages/Complexes/ChainComplexMap.m2:104]: Yes, it's the same commit, so once one of them is merged, the duplicate commit will be skipped in the other PR.

File: M2/Macaulay2/packages/Complexes/ChainComplex.m2
Comment (mahrud) [reply M2/Macaulay2/packages/Complexes/ChainComplex.m2:1165]: Thanks!

**Comment by mahrud** [comment]
> Can this pull request be decoupled from the changes to `basis`?

It is decoupled. I don't remember exactly why, but the changes to basis stemmed from fixing things in Varieties, which is why this PR is on top of that one.

> Can the changes to quotient modules and matrices to extracted?

I tried to do this a while ago but I'm afraid it was very complicated (random tests kept failing in one package or another when I tried to separate them) and I gave up, so I don't want to do it. If you or someone else wants to give it an attempt you're welcome to.

File: M2/Macaulay2/m2/matrix2.m2
Comment (mahrud) [reply M2/Macaulay2/m2/matrix2.m2:343]: I tried to do this and failed. I don't think it's impossible, but I decided it wasn't worth the time and effort to essentially redo everything in a different order.

File: M2/Macaulay2/m2/basis.m2
Comment (mahrud) [reply M2/Macaulay2/m2/basis.m2:240]: Yes, but this PR is on top of that because of actual dependencies (but it is _not_ on top of the `part` PR).

File: M2/Macaulay2/packages/Complexes/CartanEilenberg.m2
Comment (mahrud) [reply M2/Macaulay2/packages/Complexes/CartanEilenberg.m2:418]: Oh I didn't realize this file (which isn't actually loaded) was put in Complexes. I will exclude the commit that adds this for now.

File: M2/Macaulay2/m2/modules2.m2
Comment (mahrud) [reply M2/Macaulay2/m2/modules2.m2:355]: They are tested in `Varieties`, but the documentation is only a stump.

**Comment by mahrud** [comment]
Okay, updates:
- Opened #3550 for quotient changes (though this PR is still on top of it), I think the errors I couldn't fix last time were SumsOfSquares errors, I don't remember.
- Fixed the error message for non-standard graded rings
- Removed the Cartan-Eilenberg commit (it is safe on another branch for later)

File: M2/Macaulay2/packages/Varieties/backports.m2
Comment (mikestillman) [inline M2/Macaulay2/packages/Varieties/backports.m2:1]: What changes are in this file?

**Review by mikestillman — COMMENTED** [review]
Once @ggsmith is happy with this, it looks fine to me, with the caveat that I'm not so convinced that the breaking change to basis is worth the breaking.  I think `basis` needs a complete revamping at some point, with more expressive names for what it is doing.  But other than that, this all looks good.  I did not go into all the Varieties code in detail though.  I am happy to see complexes of sheaves though!

File: M2/Macaulay2/packages/Varieties/backports.m2
Comment (mahrud) [reply M2/Macaulay2/packages/Varieties/backports.m2:1]: Ah sorry this was used during the workshop, should have been removed in https://github.com/Macaulay2/M2/commit/d90df86a02f09eb22644644af8db8fa932949579. I'll do it in the next push.

**Comment by mahrud** [comment]
I have no idea why this failed after I rebased it, and the errors are in NAGtypes of all places!

**Comment by mahrud** [comment]
Figured it out, broken because of https://github.com/Macaulay2/M2/pull/3494#issuecomment-2451082376.

**Comment by mahrud** [comment]
Argh this isn't merged, I just made a mistake. I'll reopen a new one.

**Pull request merged**
